Add bruteforce, DISE, optisim dissimilarity methods

alnaba1 opened the pull request:
Added dissimilarity algorithm prototypes for brute force, directed sphere exclusion, and optisim. I plan to further optimize the functions for run-time execution and edit minor parts of the functionality. @fwmeng88 Please let me know if you have any feedback so far.

fwmeng88 commented:
Thank you for the nice contribution!

 For your PR, it's in draft status, is it ready for review? I am a little bit confused. Thanks.
Also, can you fix the problem listed in https://github.com/theochem/DiverseSelector/runs/5630230190?check_suite_focus=true#step:7:332-341? Thank you.

fwmeng88 commented:
Can you fix the failed test in https://github.com/theochem/DiverseSelector/runs/5638029542?check_suite_focus=true? Thanks. @alnaba1

alnaba1 commented:
Thank you @fwmeng88, I had my PR in draft status as I wanted to rewrite some portions of it to make it more efficient by avoiding the use of pairwise_dist for the cheaper methods, I will be making a commit tonight with the proper changes. The reason the Tox failed is because I removed the `select` method which is required according to the abstract class; I instead replaced the `select` method with my three dissimilarity methods `brutestrength`, `sphereexclusion` and `optisim` as I thought it would be best to have each algorithm in a separate method. Perhaps I should add the `select` method back and make it call the dissimilarity algorithm selected by the user? Please let me know what you think, thank you.

fwmeng88 commented:
The failing testing has been fixed. Thanks.

Can you also fix some linter problems from line 390-line 420, https://github.com/theochem/DiverseSelector/runs/5668133937?check_suite_focus=true. Thank you.

@alnaba1

fwmeng88 commented:
It seems that there are some conflicts that need to be addressed before we can merge this PR. In doing so, I would suggest backup your current version of `DiverseSelector` and trying to address the problem with 

```bash
git checkout main
# suppose you have added theochem as upstream
git pull theochem main
git checkout dissimilarity
git rebase main
# fix the conflicts according to the hints by git
```

Please don't hesitate to contact me if you have problems with this. Thanks. @alnaba1

fwmeng88 commented:
I get confused by your operation with git. Why do you merge branch `main` branch into the `dissimilarity` branch? Do you mean rebasing?

Also, I would guess you still have some tests files locally? 
Thanks for keep working on this. @alnaba1

alnaba1 commented:
Hi @fwmeng88, I tried to do rebasing locally but it wasn't showing the changes from my latest commit for some reason so I ended up fixing the conflicts on GitHub (was easy since they were so minor), hence the automated commit message.

I fixed linters and edited the structure of the `select` method to better align with how the package is used. I tried creating tests for `optisim` and `sphereexclusion` but ran into issues with `self.features_norm` being undefined, causing the algorithms to raise an error:
```bash
    def optisim(selected=None, n_selected=self.num_selected, k=self.k, r=self.r, recycling=None):
        if selected is None:
            selected = [self.starting_idx]
            return optisim(selected, n_selected, k, r, recycling)

        if len(selected) >= n_selected:
            return selected

        if recycling is None:
            recycling = []

>       candidates = np.delete(np.arange(0, len(self.features_norm)), selected + recycling)
E       AttributeError: 'DissimilaritySelection' object has no attribute 'features_norm'
```
I tried to use `self.features` in the script instead but it was set to None:
```bash
    def optisim(selected=None, n_selected=self.num_selected, k=self.k, r=self.r, recycling=None):
        if selected is None:
            selected = [self.starting_idx]
            return optisim(selected, n_selected, k, r, recycling)

        if len(selected) >= n_selected:
            return selected

        if recycling is None:
            recycling = []

>       candidates = np.delete(np.arange(0, len(self.features)), selected + recycling)
E       TypeError: object of type 'NoneType' has no len()
```

Is there an issue with `self.features` and `self.features_norm`? Not sure if I need to run a method in order for them to be assigned to proper values. This is the setup that I had for the test (very similar to current test as I just changed the select method for a quick test):
```bash
    def test_minmax_selector1():
        """Testing the MinMax selection algorithm with predefined starting point."""
        syn_data, _ = make_blobs(n_samples=100, n_features=2, centers=3, random_state=42)
        arr_dist = pairwise_distances(syn_data)
        model = DissimilaritySelection(num_selected=10, arr_dist=arr_dist, random_seed=42, method="maxmin")
        model.starting_idx = 0
>       selected = model.select("optisim")
```

alnaba1 commented:
@fwmeng88 Apologies, misunderstood how rebasing worked, will do it shortly on local and force push. Please let me know what you think of the issues regarding `self.features` and `self.features_norm` that I mentioned above.

fwmeng88 commented:
I will try to take a closer look at the `self.features` and `self.features_norm` tomorrow and get it back to you. 
@alnaba1

fwmeng88 commented on DiverseSelector/dissimilarity_based.py:
```diff
@@ -44,13 +44,18 @@ def __init__(self,
+                 k=2,
                  **kwargs,
                  ):
         """Base class for dissimilarity based subset selection."""
```
We will need detailed documentation here.

fwmeng88 commented on DiverseSelector/dissimilarity_based.py:
```diff
@@ -87,28 +92,126 @@ def compute_diversity(self):
-        if selected is None:
-            selected = [self.starting_idx]
-            return self.select(selected)
+        def brutestrength(selected=None, n_selected=self.num_selected, method=self.method):
```
Can you add documentation and a few notes for the sub-functions? You can put the notes in the *Notes* section following `numpy` documentation style.

fwmeng88 reviewed (changes requested):
Nice work. Only a few minor comments.

Also, do you have testing files on your local machine? Thanks.

fwmeng88 commented:
This is a big improvement! Thanks.
